src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,745 @@
+        if (version == 1)
+        {
+          inStream >> recordNumMinusLearnIteration_
+                   >> recordNumMinusLearnIterationSet_;
+          inStream >> numIterationHistory;
+          for (UInt i = 0; i < numIterationHistory; ++i)
+          {
+            inStream >> curIterationNum;
+            iterationNumHistory_.push_back(curIterationNum);
+          }
+        } else {
+          recordNumMinusLearnIterationSet_ = false;
+        }
+
+        // Load the prediction steps.
+        UInt size;
+        UInt step;
+        inStream >> size;
+        for (UInt i = 0; i < size; ++i)
+        {
+          inStream >> step;
+          steps_.push_back(step);
+        }
+
+        // Load the input pattern history.
+        inStream >> size;
+        UInt vSize;
+        for (UInt i = 0; i < size; ++i)
+        {
+          inStream >> vSize;
+          patternNZHistory_.emplace_back(vSize);
+          for (UInt j = 0; j < vSize; ++j)
+          {
+            inStream >> patternNZHistory_[i][j];
+          }
+          if (version == 0)
+          {
+            iterationNumHistory_.push_back(
+                learnIteration_ - (size - i));
+          }
+        }
+
+        // Load weight matrix.
+        UInt numSteps;
+        inStream >> numSteps;
+        for (UInt s = 0; s < numSteps; ++s)
+        {
+          inStream >> step;
+          // Insert the step to initialize the weight matrix
+          weightMatrix_[step] = Matrix(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+          for (UInt i = 0; i <= maxInputIdx_; ++i)
+          {
+            for (UInt j = 0; j <= maxBucketIdx_; ++j)
+            {
+              inStream >> weightMatrix_[step].at(i, j);
+            }
+          }
+        }
+
+        // Load the actual values for each bucket.
+        UInt numBuckets;
+        Real64 actualValue;
+        bool actualValueSet;
+        inStream >> numBuckets;
+        for (UInt i = 0; i < numBuckets; ++i)
+        {
+          inStream >> actualValue;
+          actualValues_.push_back(actualValue);
+          inStream >> actualValueSet;
+          actualValuesSet_.push_back(actualValueSet);
+        }
+
+        // Check for the end marker.
+        inStream >> marker;
+        NTA_CHECK(marker == "~SDRClassifier");
+
+        // Update the version number.
+        version_ = Version;
+      }
+
+      void SDRClassifier::write(SdrClassifierProto::Builder& proto) const 
+      {
+        auto stepsProto = proto.initSteps(steps_.size());
+        for (UInt i = 0; i < steps_.size(); i++)
+        {
+          stepsProto.set(i, steps_[i]);
+        }
+
+        proto.setAlpha(alpha_);
+        proto.setActValueAlpha(actValueAlpha_);
+        proto.setLearnIteration(learnIteration_);
+        proto.setRecordNumMinusLearnIteration(recordNumMinusLearnIteration_);
+        proto.setRecordNumMinusLearnIterationSet(
+            recordNumMinusLearnIterationSet_);
+        proto.setMaxSteps(maxSteps_);
```

Comment:
Oh, good point, I misread the code. The Python `deque` automatically maintains its max size, while the C++ doesn't, so we need to store the max size. Ok, this looks good (though it's definitely a bug in the Python code that it doesn't serialize the `maxSize`)